lib/instrumentation/span.js

```diff
@@ -137,8 +212,86 @@ Span.prototype.setHttpContext = function (context) {
   this._http = Object.assign(this._http || {}, context)
 }
 
-Span.prototype.setDestinationContext = function (context) {
-  this._destination = Object.assign(this._destination || {}, context)
+/**
+ * This is deprecated and will be dropped in a future version. This was always
+ * an internal method, but possibly used by enterprising users of manual
+ * instrumentation.
+ *
+ * @deprecated Users should use the public `setServiceTarget()`.
+ *    Internal APM agent code should use `_setDestinationContext()`.
+ */
+Span.prototype.setDestinationContext = function (destCtx) {
+  process.emitWarning(
```

Comment:
I don't know how necessary this warning is since `setDestinationContext` was never documented as a part of our public API. This is another example of us not being clear about what is and isn't a public API.  

By emitting this warning we seem to be saying this _was_ a part of our public API.

However, by not documenting it we were saying it's not part of our public API. 

By not having a CHANGELOG entry for this change we seem to be saying it was never a part of our public API.

It we want to start emitting these sorts of warnings when we deprecate something I don't object -- I'm just ring the "if we're going to be this fiddly we should come up with some rules" rather than making adhoc decisions about how we're doing things on a PR by PR basis.

Reply:
> I don't know how necessary this warning is

My goal in adding this warning is to *attempt* to help the accidental user of this never-public method to (a) notice and (b) have time to migrate to the new, public `.setServiceTarget()`.

While these `<span>.set<Something>Context()` methods have never been part of our public API, they sure look appealing for those attempting to do manual instrumentation:
- Those users might not pay attention to the TypeScript-y `index.d.ts`,
- those users might not rely solely on the API docs for what APIs they can use,
- there *aren't* any public APIs to use to set some of these context attributes on spans to access features like the "Service Map" and "Dependencies" in APM UI, and
- that these methods are not prefixed with an underscore **hints** that perhaps they are fair game to use.

On that last point, perhaps I am showing my [Python heritage in using a leading underscore as a hint that an attribute is non-public](https://peps.python.org/pep-0008/#method-names-and-instance-variables). Though that leading-underscore practice is used frequently in Node.js-core code as well.

Also, I feel some duty to the customer to whom I offered this code to use for manual instrumentation of Oracle (which we don't support yet): https://gist.github.com/trentm/8c8fcecbec1a99ce0fbb415ef87ae2db
That code uses `.setDbContext()` and `.setDestinationContext()`.

> By not having a CHANGELOG entry for this change we seem to be saying it was never a part of our public API.

I could certainly add a message to the changelog about this. I'll do that. I had been intending to add the following to the commit message (which is already in the PR description) -- but in the changelog is good too:

```
The not-public-but-available-because-JavaScript `span.setDestinationContext` has been deprecated (using it will `process.emitWarning()`) and replaced with an internal `span._setDestinationContext()`.
```


> By emitting this warning we seem to be saying this _was_ a part of our public API.

I'm changing to warning message to the following to attempt to alleviate that concern slightly:

```
'<span>.setDestinationContext() was never a public API and will be removed, use <span>.setServiceTarget().',
```

> "if we're going to be this fiddly we should come up with some rules" rather than making adhoc decisions 

Sounds good. I'll bring up a discussion separately. Perhaps it can lead to developer guide content in DEVELOPMENT.md